pkg/network/server_test.go

```diff
@@ -1,17 +1,176 @@
+func (f *fakeConsensus) OnPayload(p *consensus.Payload)               { f.payloads = append(f.payloads, p) }
+func (f *fakeConsensus) OnTransaction(tx *transaction.Transaction)    { f.txs = append(f.txs, tx) }
+func (f *fakeConsensus) GetPayload(h util.Uint256) *consensus.Payload { panic("implement me") }
+
+func TestNewServer(t *testing.T) {
+	bc := &testChain{}
+	s, err := newServerFromConstructors(ServerConfig{}, bc, nil, newFakeTransp, newFakeConsensus, newTestDiscovery)
+	require.Error(t, err)
+
+	t.Run("set defaults", func(t *testing.T) {
+		s = newTestServer(t, ServerConfig{MinPeers: -1})
+		defer s.discovery.Close()
+
+		require.True(t, s.ID() != 0)
+		require.Equal(t, defaultMinPeers, s.ServerConfig.MinPeers)
+		require.Equal(t, defaultMaxPeers, s.ServerConfig.MaxPeers)
+		require.Equal(t, defaultAttemptConnPeers, s.ServerConfig.AttemptConnPeers)
+	})
+	t.Run("don't defaults", func(t *testing.T) {
+		cfg := ServerConfig{
+			MinPeers:         1,
+			MaxPeers:         2,
+			AttemptConnPeers: 3,
+		}
+		s = newTestServer(t, cfg)
+		defer s.discovery.Close()
+
+		require.True(t, s.ID() != 0)
+		require.Equal(t, 1, s.ServerConfig.MinPeers)
+		require.Equal(t, 2, s.ServerConfig.MaxPeers)
+		require.Equal(t, 3, s.ServerConfig.AttemptConnPeers)
+	})
+	t.Run("consensus error is not dropped", func(t *testing.T) {
+		errConsensus := errors.New("can't create consensus")
+		_, err = newServerFromConstructors(ServerConfig{MinPeers: -1}, bc, zaptest.NewLogger(t), newFakeTransp,
+			func(consensus.Config) (consensus.Service, error) { return nil, errConsensus },
+			newTestDiscovery)
+		require.True(t, errors.Is(err, errConsensus), "got: %#v", err)
+	})
+}
+
+func TestServerStartAndShutdown(t *testing.T) {
+	t.Run("no consensus", func(t *testing.T) {
+		s := newTestServer(t, ServerConfig{})
+
+		ch := make(chan error)
+		go s.Start(ch)
+		p := newLocalPeer(t, s)
+		s.register <- p
+		s.register <- p // wait to ensure previous send was processed
+
+		assert.True(t, s.transport.(*fakeTransp).started.Load())
+		assert.False(t, s.consensus.(*fakeConsensus).started.Load())
+
+		s.Shutdown()
+		<-ch
+
+		require.True(t, s.transport.(*fakeTransp).closed.Load())
+		require.False(t, s.consensus.(*fakeConsensus).stopped.Load())
+		err, ok := p.droppedWith.Load().(error)
+		require.True(t, ok)
+		require.True(t, errors.Is(err, errServerShutdown))
+	})
+	t.Run("with consensus", func(t *testing.T) {
+		s := newTestServer(t, ServerConfig{Wallet: new(config.Wallet)})
+
+		ch := make(chan error)
+		go s.Start(ch)
+		p := newLocalPeer(t, s)
+		s.register <- p
+
+		assert.True(t, s.consensus.(*fakeConsensus).started.Load())
+
+		s.Shutdown()
+		<-ch
+
+		require.True(t, s.consensus.(*fakeConsensus).stopped.Load())
+	})
+}
+
+func TestServerRegisterPeer(t *testing.T) {
+	const peerCount = 3
+
+	s := newTestServer(t, ServerConfig{MaxPeers: 2})
+	fakeDrop := peerDrop{newLocalPeer(t, s), nil} // fake drop is used to ensure previous channel send has finished
```

Comment:
You can `require.Eventually` instead of that.

Reply:
Fixed.